### PR TITLE
Add Index Checker annotations in math package

### PR DESCRIPTION
**Pull request by NhatDinh**
@kelloggm @panacekcz Please review and let me know if any changes are needed. Thanks guys.

### CODE REVIEW

File: guava/pom.xml [hunk 1]
```diff
@@ -22,7 +22,7 @@
     <!-- Index checker warnings are disabled by default, because not all code is annotated by index checker annotations. -->
     <checkerframework.suppress>index</checkerframework.suppress>
     <!-- Packages under com.google.common, that will be additionally checked by the Index Checker -->
-    <checkerframework.index.packages>base|primitives</checkerframework.index.packages>
+    <checkerframework.index.packages>base|primitives|escape</checkerframework.index.packages>
```
Comment (panacekcz) [inline guava/pom.xml]: Enable checking the math package.

File: guava/src/com/google/common/math/BigIntegerMath.java [hunk 2]
```diff
@@ -332,7 +337,11 @@ public static BigInteger divide(BigInteger p, BigInteger q, RoundingMode mode) {
    *
    * @throws IllegalArgumentException if {@code n < 0}
    */
-  public static BigInteger factorial(int n) {
+  @SuppressWarnings(value = {"lowerbound:assignment.type.incompatible",// line 350: Since `n` is required to be non negative
+          // IntMath.divide() will also return a non negative value
+          //"lowerbound:assignment.type.incompatible"//line 356: since `factorials` length is
```
Comment (panacekcz) [inline guava/src/com/google/common/math/BigIntegerMath.java]: Should this line be removed?

File: guava/src/com/google/common/math/IntMath.java [hunk 3]
```diff
@@ -16,9 +16,9 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
-import static com.google.common.math.MathPreconditions.checkNoOverflow;
 import static com.google.common.math.MathPreconditions.checkNonNegative;
 import static com.google.common.math.MathPreconditions.checkPositive;
+import static com.google.common.math.MathPreconditions.checkNoOverflow;
```
Comment (panacekcz) [inline guava/src/com/google/common/math/IntMath.java]: Why move import?

File: guava/src/com/google/common/math/IntMath.java [hunk 4]
```diff
@@ -99,8 +105,9 @@ public static boolean isPowerOfTwo(int x) {
    * a signed int. The implementation is branch-free, and benchmarks suggest it is measurably (if
    * narrowly) faster than the straightforward ternary expression.
    */
+  @SuppressWarnings("return.type.incompatible")// lessThanBranchFree() is specified to return 1 if x < y and 0 otherwise
```
Comment (panacekcz) [inline guava/src/com/google/common/math/IntMath.java]: Can you add prefix `value:`?

Comment (panacekcz) [reply guava/src/com/google/common/math/IntMath.java]: The comment on a suppressed warning should explain why the code is correct. No need to repeat the specification which is right before.

File: guava/src/com/google/common/math/IntMath.java [hunk 5]
```diff
@@ -155,10 +162,14 @@ public static int log2(int x, RoundingMode mode) {
    *     is not a power of ten
    */
   @GwtIncompatible // need BigIntegerMath to adequately test
-  @SuppressWarnings("fallthrough")
-  public static int log10(int x, RoundingMode mode) {
+  @SuppressWarnings({"fallthrough",
+          "lowerbound:assignment.type.incompatible",/* line 172: only time when logFloor = log10Floor(x) is negative is
+          when `x` is 0, and that can't be because log10() only takes in positive `x` */
+          "upperbound:assignment.type.incompatible"/* since all elements in `maxLog10ForLeadingZeros` array
+          is @LTLengthOf("powersOf10","halfPowersOf10"), log10Floor() always return indexed values for `powersOf10` and `halfPowersOf10` */})
```
Comment (panacekcz) [inline guava/src/com/google/common/math/IntMath.java]: If that is true, then why not annotate the return type of `log10Floor` to be `@IndexFor(value = {"powersOf10", "halfPowersOf10"})`? It is like that in `LongMath`.

File: guava/src/com/google/common/math/LongMath.java [hunk 6]
```diff
@@ -102,8 +108,10 @@ public static boolean isPowerOfTwo(long x) {
    * signed long. The implementation is branch-free, and benchmarks suggest it is measurably faster
    * than the straightforward ternary expression.
    */
+
+  @SuppressWarnings("return.type.incompatible")// lessThanBranchFree() is specified to return 1 if x < y and 0 otherwise
```
Comment (panacekcz) [inline guava/src/com/google/common/math/LongMath.java]: Can you add prefix `value:`?

Comment (panacekcz) [reply guava/src/com/google/common/math/LongMath.java]: The comment on a suppressed warning should explain why the code is correct. No need to repeat the specification which is right before.

File: guava/src/com/google/common/math/Quantiles.java [hunk 7]
```diff
@@ -493,6 +520,9 @@ private static void checkIndex(int index, int scale) {
     return doubles;
   }
 
+  public static void main(String args[]) {
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Remove this code.

File: guava/src/com/google/common/math/Quantiles.java [hunk 8]
```diff
@@ -603,22 +639,26 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
...
     // ...then recursively perform the selections in the range below...
-    int requiredBelow = requiredChosen - 1;
+    int requiredBelow = requiredChosen - 1;// (1)
     while (requiredBelow >= requiredFrom && allRequired[requiredBelow] == required) {
       requiredBelow--; // skip duplicates of required in the range below
     }
     if (requiredBelow >= requiredFrom) {
-      selectAllInPlace(allRequired, requiredFrom, requiredBelow, array, from, required - 1);
+      selectAllInPlace(allRequired, requiredFrom, requiredBelow, array, from, required - 1);// (1)
+      System.out.println("required - 1: " + required);
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Remove this code.

File: guava/src/com/google/common/math/LongMath.java [hunk 9]
```diff
@@ -102,8 +108,10 @@ public static boolean isPowerOfTwo(long x) {
    * signed long. The implementation is branch-free, and benchmarks suggest it is measurably faster
    * than the straightforward ternary expression.
    */
+
```
Comment (panacekcz) [inline guava/src/com/google/common/math/LongMath.java]: Why add empty line?

**Comment by kelloggm** [comment]
The target of this PR should be `NhatDinh:cleanup-index-escape` instead of `typetools:master` so that only the changes in the `math` package are actually reviewed. We can retarget the pull request to `typetools:master` once that PR is merged.

**Comment by NhatDinh** [comment]
@panacekcz @kelloggm 
Please take a look and let me know if any changes are needed.

File: guava/src/com/google/common/math/Quantiles.java [hunk 10]
```diff
@@ -551,18 +583,20 @@ private static void selectInPlace(int required, double[] array, int from, int to
    * equal to the value at {@code ret} and the values with indexes in ({@code ret}, {@code to}] are
    * greater than or equal to that.
    */
-  private static int partition(double[] array, int from, int to) {
+  @SuppressWarnings("lowerbound:compound.assignment.type.incompatible")// (1): Lowest possible `from` value is 0, Since for loop init at `i = to`, `to` needs to be at least at 1 in order to execute for loop
+  // If `to` is at least 1, `partionPoint` can't be negative
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Why `partitionPoint` can't be negative? I don't think it is obvious.

File: guava/src/com/google/common/math/Quantiles.java [hunk 11]
```diff
@@ -603,22 +637,25 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
    * allRequired[i]} for {@code i} in the range [{@code requiredFrom}, {@code requiredTo}]. These
    * indexes must be sorted in the array and must all be in the range [{@code from}, {@code to}].
    */
+  @SuppressWarnings({"lowerbound:argument.type.incompatible",// (1): lowest `requiredFrom` value is 0, for `requiredBelow >= requiredFrom`, required is least 1
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: I don't see why `required` is at least 1. Can you explain more?

**Comment by NhatDinh** [comment]
I think the two errors in Travis are caused because the used index checker's version is not up to date.
The two errors are similar to this [integration test](https://github.com/typetools/checker-framework/pull/2022/commits/949da83602b058a86e788633bbdd4d13a18d0728) included in the merged [PR 2022](https://github.com/typetools/checker-framework/pull/2022/files)
@panacekcz

**Comment by NhatDinh** [comment]
I think this PR is ready for another round of reviews and possibly a merge.
@kelloggm @panacekcz

File: guava/src/com/google/common/math/Quantiles.java [hunk 12]
```diff
@@ -513,7 +542,11 @@ private static void checkIndex(int index, int scale) {
    * ({@code required}, {@code to}] are greater than or equal to that value. Therefore, the value at
    * {@code required} is the value which would appear at that index in the sorted dataset.
    */
-  private static void selectInPlace(int required, double[] array, int from, int to) {
+  @SuppressWarnings(value = {"lowerbound:assignment.type.incompatible",//(1): lowest possible while `to` > 0, `partitionPoint` value is 1, therefore `to` can't be negative.
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: I think `to > 0` is not relevant here.
When entering the loop, `required > from` and `from >= 0`, therefore `required >= 1`.
At the assignment, `partitionPoint >= required`, therefore `partitionPoint - 1 >= 0`.

File: guava/src/com/google/common/math/Quantiles.java [hunk 13]
```diff
@@ -513,7 +542,11 @@ private static void checkIndex(int index, int scale) {
...
+          "upperbound:assignment.type.incompatible"/* (2): highest possible `partionPoint` value is array.length / 2 or (array.length / 2) + 1.
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Really? Why?

File: guava/src/com/google/common/math/Quantiles.java [hunk 14]
```diff
@@ -551,18 +584,21 @@ private static void selectInPlace(int required, double[] array, int from, int to
    * equal to the value at {@code ret} and the values with indexes in ({@code ret}, {@code to}] are
    * greater than or equal to that.
    */
-  private static int partition(double[] array, int from, int to) {
+  @SuppressWarnings("lowerbound:compound.assignment.type.incompatible")// (1): Lowest possible `from` value is 0.
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: I'd explain it like this:
Both `partitionPoint` and `i` are initialized to the same value (`to`).
`partitionPoint` is decremented at most one more time than `i` is decremented, therefore `partitionPoint >= i-1`.
`i > from`, therefore `partitionPoint >= from`. `from` is non-negative, therefore `partitionPoint` is non-negative.

File: guava/src/com/google/common/math/Quantiles.java [hunk 15]
```diff
@@ -380,7 +397,19 @@ private ScaleAndIndexes(int scale, int[] indexes) {
      * @return an unmodifiable map of results: the keys will be the specified quantile indexes, and
      *     the values the corresponding quantile values
      */
-    public Map<Integer, Double> computeInPlace(double... dataset) {
+    @SuppressWarnings({"upperbound:assignment.type.incompatible",// indexes` array is annotated to have min length of 1( which is > 0).
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: What line does this refer to?

**Comment by NhatDinh** [comment]
I think it's ready for a merge or we're very close now.
@panacekcz

File: guava/src/com/google/common/math/Quantiles.java [hunk 16]
```diff
@@ -603,22 +638,26 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
    * allRequired[i]} for {@code i} in the range [{@code requiredFrom}, {@code requiredTo}]. These
    * indexes must be sorted in the array and must all be in the range [{@code from}, {@code to}].
    */
+  @SuppressWarnings({"lowerbound:argument.type.incompatible",/* (1): Lowest possible `requiredFrom` value is 0.
+           Although `requiredBelow` can be negative, if `requiredBelow` is not >= `requiredFrom`, `required` can't be negative */
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: This does not explain why the code is correct at all.
You'd need to show that `required` is positive on that line.
I think it relies on properties of the content of the arrays.

File: guava/src/com/google/common/math/Quantiles.java [hunk 17]
```diff
@@ -603,22 +638,26 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
...
+          "upperbound:argument.type.incompatible"/* (2): highest `requiredTo` value is `allRequired.length - 1`.
+          When `requiredAbove` <= highest `requiredTo` value, required max value is `allRequired.length - 1`*/ })
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: This does not make sense and does not explain why the code is correct.

**Comment by NhatDinh** [comment]
Please review and merge if possible.
@kelloggm @panacekcz

File: guava/src/com/google/common/math/Quantiles.java [hunk 18]
```diff
@@ -603,22 +639,30 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
    * allRequired[i]} for {@code i} in the range [{@code requiredFrom}, {@code requiredTo}]. These
    * indexes must be sorted in the array and must all be in the range [{@code from}, {@code to}].
    */
+  @SuppressWarnings({"lowerbound:argument.type.incompatible",/*(1): Since allRequired is sorted, and requiredBelow < requiredChosen,
+  allRequired[requiredBelow] <= allRequired[requiredChosen], therefore allRequired[requiredBelow] is always <= required. Based on the condition
+  of if and while, allRequired[requiredBelow] != required. Because allRequired[requiredBelow] <= required, we know that allRequired[requiredBelow] < required,
+  since allRequired[requiredBelow] >= requiredFrom >= 0, then required > 0. */
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: I don't see why `allRequired[requiredBelow] >= requiredFrom` should hold. Can you simply remove that from the equation and keep just `allRequired[requiredBelow] >= 0`?

File: guava/src/com/google/common/math/Quantiles.java [hunk 19]
```diff
@@ -603,22 +639,30 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
...
+          "upperbound:argument.type.incompatible"/*(2): Since allRequired is sorted, and requiredAbove > requiredChosen,
+  allRequired[requiredAbove] >= allRequired[requiredChosen], therefore allRequired[requiredAbove] is always >= required. Based on the condition
+  of if and while, allRequired[requiredAbove] != required. Because allRequired[requiredAbove] >= required, we know that allRequired[requiredAbove] > required,
+  since requiredTo >= allRequired[requiredAbove] > required, required + 1 < array.length. */})
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Again, I don't see why `requiredTo >= allRequired[requiredAbove]` should hold. Can you write `array.length > allRequired[requiredAbove]` instead?
Otherwise, the explanation looks good.

File: guava/src/com/google/common/math/Quantiles.java [hunk 20]
```diff
@@ -30,6 +30,9 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import org.checkerframework.checker.index.qual.*;
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Format imports according to the style guide.

File: guava/src/com/google/common/math/Quantiles.java [hunk 21]
```diff
@@ -513,7 +540,13 @@ private static void checkIndex(int index, int scale) {
    * ({@code required}, {@code to}] are greater than or equal to that value. Therefore, the value at
    * {@code required} is the value which would appear at that index in the sorted dataset.
    */
-  private static void selectInPlace(int required, double[] array, int from, int to) {
+  @SuppressWarnings(value = {"lowerbound:assignment.type.incompatible",/*(1): When entering the loop,
+          required > from and from >= 0, therefore required >= 1.
+          At the assignment, partitionPoint >= required, therefore partitionPoint - 1 >= 0. */
+          "upperbound:assignment.type.incompatible"/*(2): To and from are both < array.length. Entering the loop, from < to,
+          from < array.length - 1, therefore from = partitionPoint + 1 < array.length.
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: This doesn't explain why `partitionPoint + 1 < array.length`. Actually, it seems that it doesn't have to be always true. If `from == array.length`, then `from > to` and the loop ends, so the code is still safe.

File: guava/src/com/google/common/math/Quantiles.java [hunk 22]
```diff
@@ -551,18 +584,21 @@ private static void selectInPlace(int required, double[] array, int from, int to
    * equal to the value at {@code ret} and the values with indexes in ({@code ret}, {@code to}] are
    * greater than or equal to that.
    */
-  private static int partition(double[] array, int from, int to) {
+  @SuppressWarnings("lowerbound:compound.assignment.type.incompatible")/*(1): Both partitionPoint and i are initialized to the same value `to`.
+  partitionPoint is decremented at most one more time than i is decremented, therefore partitionPoint >= i-1.
+  i > from, therefore partitionPoint >= from. Since from is non-negative, therefore partitionPoint is non-negative. */
+  private static @IndexFor("#1") int partition( double[] array, @IndexFor("#1") int from, @IndexFor("#1") int to) {
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Remove space before `double`.

File: guava/src/com/google/common/math/Quantiles.java [hunk 23]
```diff
@@ -603,22 +639,30 @@ private static void movePivotToStartOfSlice(double[] array, int from, int to) {
...
+  of if and while, allRequired[requiredBelow] != required. Because allRequired[requiredBelow] <= required, we know that allRequired[requiredBelow] < required,
+  since allRequired[requiredBelow] >= requiredFrom >= 0, then required > 0. */
+          "upperbound:argument.type.incompatible"/*(2): Since allRequired is sorted, and requiredAbove > requiredChosen,
+  allRequired[requiredAbove] >= allRequired[requiredChosen], therefore allRequired[requiredAbove] is always >= required. Based on the condition
+  of if and while, allRequired[requiredAbove] != required. Because allRequired[requiredAbove] >= required, we know that allRequired[requiredAbove] > required,
+  since requiredTo >= allRequired[requiredAbove] > required, required + 1 < array.length. */})
   private static void selectAllInPlace(
-      int[] allRequired, int requiredFrom, int requiredTo, double[] array, int from, int to) {
+          @IndexFor("#4") int[] allRequired, @IndexFor("#1") int requiredFrom, @IndexFor("#1") int requiredTo, double[] array, @IndexFor("#4") int from, @IndexFor("#4") int to) {
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: Unnecessary whitespace change at the start of the line.

**Comment by NhatDinh** [comment]
I hope it is ready now.
@kelloggm @panacekcz

**Comment by NhatDinh** [comment]
I've removed 2 suppress warnings because of [PR #2103](https://github.com/typetools/checker-framework/pull/2103)
@kelloggm @panacekcz

File: guava/src/com/google/common/math/BigIntegerMath.java [hunk 24]
```diff
@@ -144,11 +148,12 @@ public static int log2(BigInteger x, RoundingMode mode) {
    *     is not a power of ten
    */
   @GwtIncompatible // TODO
-  @SuppressWarnings("fallthrough")
-  public static int log10(BigInteger x, RoundingMode mode) {
+  @SuppressWarnings({"fallthrough","lowerbound:argument.type.incompatible"// if x is posItive, `x.longValue()` returns positive
```
Comment (kelloggm) [inline guava/src/com/google/common/math/BigIntegerMath.java]: The important property here isn't that `x` is positive (which the checker does know). It's that if `x` fits in a `long`, then `longValue` is polymorphic over `x`'s value with respect to the lowerbound type system. I suggest changing the explanation to make that clear.

File: guava/src/com/google/common/math/DoubleMath.java [hunk 25]
```diff
@@ -38,13 +38,16 @@
...
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.common.value.qual.MinLen;
 
 /**
  * A class for arithmetic on doubles that is not covered by {@link java.lang.Math}.
  *
  * @author Louis Wasserman
  * @since 11.0
  */
+@SuppressWarnings("expression.unparsable.type.invalid")// link to issue: https://github.com/typetools/checker-framework/issues/2030
```
Comment (kelloggm) [inline guava/src/com/google/common/math/DoubleMath.java]: This issue is closed, and there was correspondence on it from Mike/Suzanne that you didn't respond to. Why is it still linked here if it's (apparently) no longer an issue?

File: guava/src/com/google/common/math/LongMath.java [hunk 26]
```diff
@@ -805,7 +828,8 @@ public static long factorial(int n) {
    *
    * @throws IllegalArgumentException if {@code n < 0}, {@code k < 0}, or {@code k > n}
    */
-  public static long binomial(int n, int k) {
+  @SuppressWarnings("lowerbound:compound.assignment.type.incompatible")// the lowest n can be is 0 in this method
```
Comment (kelloggm) [inline guava/src/com/google/common/math/LongMath.java]: This is a fact that's already conveyed by the `@NonNegative` annotation on `n`. You should explain *why* `n` cannot actually violate the invariant - the annotations already express what the invariant is.

File: guava/src/com/google/common/math/Quantiles.java [hunk 27]
```diff
@@ -199,6 +205,7 @@ public ScaleAndIndexes indexes(int... indexes) {
      *     q-quantiles; the order of the indexes is unimportant, duplicates will be ignored, and the
      *     set will be snapshotted when this method is called
      */
+    @SuppressWarnings({"value:argument.type.incompatible", "lowerbound:argument.type.incompatible"})//para `indexes` is of mutable length data structures(Collection)
```
Comment (kelloggm) [inline guava/src/com/google/common/math/Quantiles.java]: What does "para" mean here?

**Comment by NhatDinh** [comment]
Please give it another round of review.
@kelloggm @panacekcz

File: guava/src/com/google/common/math/DoubleMath.java [hunk 28]
```diff
@@ -51,6 +53,7 @@
    * This method returns a value y such that rounding y DOWN (towards zero) gives the same result as
    * rounding x according to the specified mode.
    */
+  @SuppressWarnings("expression.unparsable.type.invalid")//Link to issue: https://github.com/typetools/checker-framework/issues/2030( marked as unreproducible)
```
Comment (panacekcz) [inline guava/src/com/google/common/math/DoubleMath.java]: If #2030 issue has not been fixed, why is it closed?
Don't state that the issue was marked as unreproducible here. That should be discussed in the issue itself, this comment might get out of date and confuse people.

Comment (NhatDinh) [reply guava/src/com/google/common/math/DoubleMath.java]: I've reopened the issue again and provided a different test case
@panacekcz

File: guava/src/com/google/common/math/Quantiles.java [hunk 29]
```diff
@@ -513,7 +543,12 @@ private static void checkIndex(int index, int scale) {
    * ({@code required}, {@code to}] are greater than or equal to that value. Therefore, the value at
    * {@code required} is the value which would appear at that index in the sorted dataset.
    */
-  private static void selectInPlace(int required, double[] array, int from, int to) {
+  @SuppressWarnings(value = {"lowerbound:assignment.type.incompatible",/*(1): When entering the loop,
+          required > from and from >= 0, therefore required >= 1.
+          At the assignment, partitionPoint >= required, therefore partitionPoint - 1 >= 0. */
+          "upperbound:assignment.type.incompatible"/*(2): Entering the loop, from and to < array.length. While to > from,
+          to > partitionPoint + 1, therefore partitionPoint + 1 < array.length */})
```
Comment (panacekcz) [inline guava/src/com/google/common/math/Quantiles.java]: It seems to me that after `int partitionPoint = partition(array, from, to);`, it can be that `partitionPoint == to`, contrary to this comment. Is that right?
Maybe you mean that it holds at certain point in the program? However, that is not stated in the comment, and it is not obvious. For someone trying to understand the code, it would be just confusing.
I definitely want to avoid statements that are, or look like, a contradiction.
If you know why the code is correct, can you write explicitly all the steps you used to assert that? It may be longer, but better a longer clear explanation than a short confusing one.

File: guava/src/com/google/common/math/LongMath.java [hunk 30]
```diff
@@ -184,33 +194,46 @@ public static int log10(long x, RoundingMode mode) {
   }
 
   @GwtIncompatible // TODO
-  static int log10Floor(long x) {
+  @SuppressWarnings(value = {"lowerbound:return.type.incompatible",/* (2): `log10Floor()` return negative int value only when y = 0 and
+          `LessThanBranchFree` return 1( when x < y). Since `log10floor()` is a static method and only called by methods that take in positive `x` values.
```
Comment (panacekcz) [inline guava/src/com/google/common/math/LongMath.java]: This does not seem correct. The comment on the same warining in `IntMath.log10Floor` looks correct, so you could use it here. Although I think you can leave out the sentence about `maxLog10ForLeadingZeros`, because what you need is to show that `0 < x`, and that is implied by the `@Positive` annotation on `x`.

Comment (NhatDinh) [reply guava/src/com/google/common/math/LongMath.java]: I've replaced the description with one similar in `IntMat#log10floor()`. I still think it's important to keep the sentence about maxLog10ForLeadingZeros.

File: guava/src/com/google/common/math/Quantiles.java (hunk 29 above)
Comment (NhatDinh) [reply guava/src/com/google/common/math/Quantiles.java]: Please let me know if the new replaced comment makes better sense.

File: guava/src/com/google/common/math/DoubleMath.java (hunk 28 above)
Comment (panacekcz) [reply guava/src/com/google/common/math/DoubleMath.java]: Thanks.

**Comment by NhatDinh** [comment]
I've added the local variable in `Quantiles#selectInPlace()` as discussed.
@kelloggm @panacekcz

**Comment by panacekcz** [comment]
@kelloggm, do you want to review this again? Otherwise, I think I can merge.
